File: typescript/src/agents/facades/openai.ts
/**
 * OpenAI-shaped facade — ``client.chat.completions.create(...)``.
 *
 * Delegates the actual run to a private :class:`CompresrEngine` and remaps the
 * returned :class:`NormalizedResult` to an OpenAI ``ChatCompletion`` shape.
 *
 * Mirrors Python ``compresr/agents/facades/openai.py``.
 */
import { CompresrError } from '../../errors/index.js';
import { CompresrEngine } from '../engine.js';
import { toChatCompletion, type ChatCompletion } from '../schemas/openai.js';

export interface OpenAICompletionsCreateOptions {
  /**
   * Optional per-call model. Defaults to the model encoded on the engine's
   * ``llm`` spec (``'openai:gpt-...'``). If neither is set, the engine throws
   * ``CompresrError("missing_model")``.
   */
  model?: string;
  messages: ReadonlyArray<unknown>;
  maxTokens?: number;
  tools?: ReadonlyArray<unknown>;
  config?: Record<string, unknown>;
  temperature?: number;
  topP?: number;
  presencePenalty?: number;
  frequencyPenalty?: number;
  seed?: number;
  stop?: string[];
  logprobs?: boolean;
  topLogprobs?: number;
}

export interface OpenAICompletionsFacade {
  create(options: OpenAICompletionsCreateOptions): Promise<ChatCompletion>;
  /** Streaming not yet implemented — throws. */
  stream(options: OpenAICompletionsCreateOptions): Promise<never>;
}

export interface OpenAIChatFacade {
  completions: OpenAICompletionsFacade;
}

export function openaiChatCompletions(engine: CompresrEngine): OpenAIChatFacade {
  const completions: OpenAICompletionsFacade = {
    async create(options: OpenAICompletionsCreateOptions): Promise<ChatCompletion> {
      const runOpts: Parameters<typeof engine.run>[0] = {
        messages: options.messages,
        tools: options.tools ?? [],
        maxTokens: options.maxTokens ?? 4096,
      };
      if (options.model !== undefined) runOpts.model = options.model;
      if (options.config !== undefined) runOpts.config = options.config;
      if (options.temperature !== undefined) runOpts.temperature = options.temperature;
      if (options.topP !== undefined) runOpts.topP = options.topP;
      if (options.presencePenalty !== undefined) {
        runOpts.presencePenalty = options.presencePenalty;
      }
      if (options.frequencyPenalty !== undefined) {
        runOpts.frequencyPenalty = options.frequencyPenalty;
      }
      if (options.seed !== undefined) runOpts.seed = options.seed;
      if (options.stop !== undefined) runOpts.stop = options.stop;
      if (options.logprobs !== undefined) runOpts.logprobs = options.logprobs;
      if (options.topLogprobs !== undefined) {
        runOpts.topLogprobs = options.topLogprobs;
      }
      const result = await engine.run(runOpts);
      const responseModel = options.model ?? engine.defaultModelName ?? '';
      return toChatCompletion(result, { model: responseModel });
    },
    stream(_options: OpenAICompletionsCreateOptions): Promise<never> {
      return Promise.reject(
        new CompresrError(
          'streaming not yet implemented — Phase 2 work item',
          'not_implemented'
        )
      );
    },
  };
  return { completions };
}

File: python/compresr/agents/facades/openai.py
"""OpenAI-shaped facade — ``client.chat.completions.create(...)``.

Delegates the actual run to a private :class:`_Engine` and remaps the
returned :class:`NormalizedResult` to an OpenAI ``ChatCompletion`` shape.

Note: OpenAI puts the system prompt inside the ``messages`` list (with
``role="system"``). We pass ``messages`` through unchanged.
"""

from __future__ import annotations

from typing import Any, Optional

from ..schemas.openai import ChatCompletion, to_chat_completion


def _forward(kw: dict, **maybe: Any) -> dict:
    """Return ``kw`` with each non-``None`` ``maybe`` value added."""
    for key, value in maybe.items():
        if value is not None:
            kw[key] = value
    return kw


class _OpenAICompletions:
    """``client.chat.completions`` surface — exposes ``.create`` / ``.acreate``."""

    def __init__(self, engine: Any) -> None:
        self._engine = engine

    def create(
        self,
        *,
        model: Optional[str] = None,
        messages: list,
        max_tokens: int = 4096,
        tools: Optional[list] = None,
        temperature: Optional[float] = None,
        top_p: Optional[float] = None,
        presence_penalty: Optional[float] = None,
        frequency_penalty: Optional[float] = None,
        seed: Optional[int] = None,
        stop: Optional[Any] = None,
        logprobs: Optional[bool] = None,
        top_logprobs: Optional[int] = None,
        **kw: Any,
    ) -> ChatCompletion:
        forwarded = _forward(
            dict(kw),
            temperature=temperature,
            top_p=top_p,
            presence_penalty=presence_penalty,
            frequency_penalty=frequency_penalty,
            seed=seed,
            stop=stop,
            logprobs=logprobs,
            top_logprobs=top_logprobs,
        )
        result = self._engine.run(
            messages=messages,
            tools=tuple(tools or ()),
            max_tokens=max_tokens,
            model=model,
            **forwarded,
        )
        effective_model = model or self._engine.default_model_name
        return to_chat_completion(result, model=effective_model)

    async def acreate(
        self,
        *,
        model: Optional[str] = None,
        messages: list,
        max_tokens: int = 4096,
        tools: Optional[list] = None,
        temperature: Optional[float] = None,
        top_p: Optional[float] = None,
        presence_penalty: Optional[float] = None,
        frequency_penalty: Optional[float] = None,
        seed: Optional[int] = None,
        stop: Optional[Any] = None,
        logprobs: Optional[bool] = None,
        top_logprobs: Optional[int] = None,
        **kw: Any,
    ) -> ChatCompletion:
        forwarded = _forward(
            dict(kw),
            temperature=temperature,
            top_p=top_p,
            presence_penalty=presence_penalty,
            frequency_penalty=frequency_penalty,
            seed=seed,
            stop=stop,
            logprobs=logprobs,
            top_logprobs=top_logprobs,
        )
        result = await self._engine.arun(
            messages=messages,
            tools=tuple(tools or ()),
            max_tokens=max_tokens,
            model=model,
            **forwarded,
        )
        effective_model = model or self._engine.default_model_name
        return to_chat_completion(result, model=effective_model)


class _OpenAIChat:
    """``client.chat`` surface — exposes ``.completions``."""

    def __init__(self, engine: Any) -> None:
        self.completions = _OpenAICompletions(engine)


class _OpenAI:
    """``client`` surface — exposes ``.chat``."""

    def __init__(self, engine: Any) -> None:
        self.chat = _OpenAIChat(engine)


__all__ = ["_OpenAI", "_OpenAIChat", "_OpenAICompletions"]
